pin/gc/gc.go

```diff
@@ -35,144 +54,159 @@ type Result struct {
 //
 func GC(ctx context.Context, bs bstore.GCBlockstore, ls dag.LinkService, pn pin.Pinner, bestEffortRoots []*cid.Cid) <-chan Result {
 
-	elock := log.EventBegin(ctx, "GC.lockWait")
-	unlocker := bs.GCLock()
-	elock.Done()
-	elock = log.EventBegin(ctx, "GC.locked")
-	emark := log.EventBegin(ctx, "GC.mark")
-
 	ls = ls.GetOfflineLinkService()
-
 	output := make(chan Result, 128)
+	tri := newTriset()
+
+	emark := log.EventBegin(ctx, "GC.mark")
+
+	err := pn.Flush()
+	if err != nil {
+		output <- Result{Error: err}
+		close(output)
+		return output
+	}
+	addRoots(tri, pn, bestEffortRoots)
 
 	go func() {
 		defer close(output)
-		defer unlocker.Unlock()
-		defer elock.Done()
 
-		gcs, err := ColoredSet(ctx, pn, ls, bestEffortRoots, output)
-		if err != nil {
-			output <- Result{Error: err}
+		bestEffortGetLinks := func(ctx context.Context, cid *cid.Cid) ([]*node.Link, error) {
+			links, err := ls.GetLinks(ctx, cid)
+			if err != nil && err != dag.ErrNotFound {
+				return nil, &CannotFetchLinksError{cid, err}
+			}
+			return links, nil
+		}
+
+		getLinks := func(ctx context.Context, cid *cid.Cid) ([]*node.Link, error) {
+			links, err := ls.GetLinks(ctx, cid)
+			if err != nil {
+				return nil, &CannotFetchLinksError{cid, err}
+			}
+			return links, nil
+		}
+
+		var criticalError error
+		defer func() {
+			if criticalError != nil {
+				output <- Result{Error: criticalError}
+			}
+		}()
+
+		// Enumerate without the lock
+		for {
+			finished, err := tri.EnumerateStep(ctx, bestEffortGetLinks, getLinks)
+			if err != nil {
+				output <- Result{Error: err}
+				criticalError = ErrCannotFetchAllLinks
```

Comment:
we dont want to halt here? Whats the point in continuing?

Reply:
To collect all errors like that.

Reply:
I would add that this is how I modified the GC to work in https://github.com/ipfs/go-ipfs/pull/3712.  It make it easier for the user to figure out and fix the problem in one go.